File: iOS/LiquorApp/AppServicesSyncManager.swift
import CouchbaseLiteSwift
import Foundation
import Combine

// MARK: - App Services Sync State
struct AppServicesSyncState {
    var isConnected: Bool = false
    var status: String = "Disconnected"
    var lastSyncTime: Date? = nil
    var documentsInSync: Int = 0
    var progress: Float = 0.0
    var error: String? = nil
    var totalDocuments: Int = 0
    var documentsCompleted: Int = 0
}

// MARK: - App Services Sync Manager
class AppServicesSyncManager: ObservableObject {
    
    // MARK: - Published Properties
    @Published var syncState = AppServicesSyncState()
    @Published var isEnabled: Bool = false
    
    // MARK: - Configuration (from AppConfig)
    private let syncGatewayURL = AppConfig.syncGatewayURL
    private let username = AppConfig.username
    private let password = AppConfig.password
    
    // MARK: - Core Components
    private var database: Database
    private var replicator: Replicator?
    private var replicatorChangeToken: ListenerToken?
    private let collectionName = AppConfig.collectionName
    
    // MARK: - Sync Control
    private var isSyncActive = false
    private let syncQueue = DispatchQueue(label: "com.liquorapp.appsync", qos: .background)
    
    // MARK: - Init
    init(database: Database) {
        self.database = database
        print("🌐 AppServicesSyncManager initialized with database: \(database.name)")
        setupAppServicesSync()
    }
    
    // MARK: - Setup Methods
    private func setupAppServicesSync() {
        print("🔧 Setting up App Services sync configuration...")
        print("🔧 Scope: \(AppConfig.scopeName), Collection: \(collectionName)")
        
        do {
            // Ensure collection exists (using scope from AppConfig to match Capella structure)
            let collection = try database.collection(name: collectionName, scope: AppConfig.scopeName) 
                ?? database.createCollection(name: collectionName, scope: AppConfig.scopeName)
            
            // Create target endpoint
            guard let url = URL(string: syncGatewayURL) else {
                throw NSError(domain: "Invalid sync gateway URL", code: -1)
            }
            
            let target = URLEndpoint(url: url)
            var config = ReplicatorConfiguration(target: target)
            
            // Configure authentication
            config.authenticator = BasicAuthenticator(username: username, password: password)
            
            // Configure replication type and behavior
            config.replicatorType = .pushAndPull
            config.continuous = true
            config.enableAutoPurge = false
            config.heartbeat = 60 // seconds
            config.maxAttempts = 10
            config.maxAttemptWaitTime = 300 // 5 minutes
            config.allowReplicatingInBackground = true
            
            // Configure collections with CRDT conflict resolver
            var collectionConfig = CollectionConfiguration()
            collectionConfig.conflictResolver = LiquorCRDTConflictResolver.shared
            config.addCollection(collection, config: collectionConfig)
            
            // Create replicator
            replicator = Replicator(config: config)
            
            // Add change listener
            replicatorChangeToken = replicator?.addChangeListener { [weak self] change in
                DispatchQueue.main.async {
                    self?.handleReplicationChange(change)
                }
            }
            
            print("✅ App Services sync configured successfully")
            updateSyncState { state in
                state.status = "☁️ Ready to sync"
            }
            
        } catch {
            print("❌ Failed to setup App Services sync: \(error)")
            updateSyncState { state in
                state.status = "Setup failed"
                state.error = error.localizedDescription
            }
        }
    }
    
    // MARK: - Public Sync Control Methods
    func enableAppServices() {
        guard !isEnabled else { return }
        
        print("🚀 Enabling App Services sync...")
        isEnabled = true
        startSync()
        
        updateSyncState { state in
            state.status = "☁️ Starting cloud sync..."
        }
    }
    
    func disableAppServices() {
        guard isEnabled else { return }
        
        print("🛑 Disabling App Services sync...")
        isEnabled = false
        stopSync()
        
        updateSyncState { state in
            state.status = "☁️ Cloud sync stopped"
            state.isConnected = false
        }
    }
    
    func toggleAppServices() {
        if isEnabled {
            disableAppServices()
        } else {
            enableAppServices()
        }
    }
    
    private func startSync() {
        guard let replicator = replicator, !isSyncActive else {
            print("⚠️ Cannot start sync - replicator not available or already active")
            return
        }
        
        syncQueue.async { [weak self] in
            guard let self = self else { return }
            
            print("🌐 Starting App Services replicator...")
            self.isSyncActive = true
            replicator.start()
            
            DispatchQueue.main.async {
                self.updateSyncState { state in
                    state.status = "☁️ Connecting to cloud..."
                }
            }
        }
    }
    
    private func stopSync() {
        guard let replicator = replicator, isSyncActive else { return }
        
        syncQueue.async { [weak self] in
            guard let self = self else { return }
            
            print("🛑 Stopping App Services replicator...")
            replicator.stop()
            self.isSyncActive = false
            
            DispatchQueue.main.async {
                self.updateSyncState { state in
                    state.status = "☁️ Sync stopped"
                    state.isConnected = false
                }
            }
        }
    }
    
    func resetSync() {
        print("🔄 Resetting App Services sync...")
        
        stopSync()
        
        // Reset checkpoint to force complete resync
        syncQueue.asyncAfter(deadline: .now() + 1.0) { [weak self] in
            // Note: resetCheckpoint might not be available in all Couchbase Lite versions
            // Alternative: stop and restart replicator to trigger a fresh sync
            
            DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
                self?.startSync()
            }
        }
        
        updateSyncState { state in
            state.status = "☁️ Resetting sync..."
            state.progress = 0.0
            state.error = nil
        }
    }
    
    // MARK: - Replication Event Handling
    private func handleReplicationChange(_ change: ReplicatorChange) {
        let status = change.status
        let progress = status.progress
        
        print("📊 App Services sync change: \(status.activity) - \(progress.completed)/\(progress.total)")
        
        updateSyncState { state in
            // Update connection status
            state.isConnected = (status.activity == .busy || status.activity == .idle)
            
            // Update progress
            if progress.total > 0 {
                state.progress = Float(progress.completed) / Float(progress.total)
                state.documentsCompleted = Int(progress.completed)
                state.totalDocuments = Int(progress.total)
                state.documentsInSync = Int(progress.total)
            }
            
            // Update status based on activity
            switch status.activity {
            case .connecting:
                state.status = "☁️ Connecting to cloud..."
                
            case .busy:
                state.status = "☁️ Syncing... (\(progress.completed)/\(progress.total))"
                
            case .idle:
                state.status = "☁️ Cloud sync ready"
                state.lastSyncTime = Date()
                state.progress = 1.0
                
            case .stopped:
                state.status = "☁️ Sync stopped"
                state.isConnected = false
                state.progress = 0.0
                
            case .offline:
                state.status = "☁️ Cloud offline"
                state.isConnected = false
                
            @unknown default:
                state.status = "☁️ Unknown sync state"
            }
            
            // Handle errors
            if let error = status.error {
                print("❌ App Services sync error: \(error)")
                
                // Handle different error types
                let errorCode = (error as NSError).code
                switch errorCode {
                case 11001: // Network error (based on common error codes)
                    state.status = "☁️ Network error - will retry"
                    state.error = "Network connectivity issue"
                case 11002: // Auth required
                    state.status = "☁️ Authentication failed"
                    state.error = "Invalid credentials"
                case 11003: // Forbidden
                    state.status = "☁️ Access denied"
                    state.error = "Permission denied"
                case 11004: // Not found
                    state.status = "☁️ Database not found"
                    state.error = "Remote database not found"
                default:
                    state.status = "☁️ Sync error"
                    state.error = error.localizedDescription
                }
                
                state.isConnected = false
            } else {
                state.error = nil
            }
        }
    }
    
    // MARK: - Document Operations
    func pushDocumentImmediately(_ documentId: String) {
        guard isEnabled, replicator != nil else {
            print("⚠️ Cannot push document - sync not enabled")
            return
        }
        
        print("📤 Triggering immediate push for document: \(documentId)")
        
        // The document will be automatically picked up by the continuous replicator
        // We can trigger a one-time sync to speed things up
        if !isSyncActive {
            startSync()
        }
    }
    
    func getConflictedDocuments() -> [String] {
        do {
            let collection = try database.collection(name: collectionName, scope: AppConfig.scopeName)
            let query = QueryBuilder
                .select(SelectResult.expression(Meta.id))
                .from(DataSource.collection(collection!))
                .where(Meta.revisionID.like(Expression.string("%-%"))) // Simple conflict detection
            
            let results = try query.execute()
            var conflicts: [String] = []
            
            for result in results {
                if let docId = result.string(at: 0) {
                    conflicts.append(docId)
                }
            }
            
            print("🔍 Found \(conflicts.count) conflicted documents")
            return conflicts
            
        } catch {
            print("❌ Failed to query conflicts: \(error)")
            return []
        }
    }
    
    // MARK: - Helper Methods
    private func updateSyncState(_ update: (inout AppServicesSyncState) -> Void) {
        var newState = syncState
        update(&newState)
        syncState = newState
    }
    
    // MARK: - Status Information
    func getSyncStatusSummary() -> String {
        if !isEnabled {
            return "App Services sync disabled"
        }
        
        let baseStatus = syncState.status
        
        if syncState.isConnected {
            if let lastSync = syncState.lastSyncTime {
                let formatter = DateFormatter()
                formatter.timeStyle = .short
                return "\(baseStatus) • Last: \(formatter.string(from: lastSync))"
            } else {
                return baseStatus
            }
        } else {
            return baseStatus
        }
    }
    
    func getDebugInfo() -> [String: Any] {
        return [
            "enabled": isEnabled,
            "connected": syncState.isConnected,
            "status": syncState.status,
            "progress": syncState.progress,
            "documents_in_sync": syncState.documentsInSync,
            "last_sync": syncState.lastSyncTime?.timeIntervalSince1970 ?? 0,
            "error": syncState.error ?? "none",
            "replicator_active": isSyncActive,
            "sync_gateway_url": syncGatewayURL,
            "username": username
        ]
    }
    
    // MARK: - Cleanup
    deinit {
        print("🧹 Cleaning up AppServicesSyncManager...")
        
        stopSync()
        
        if let token = replicatorChangeToken {
            replicator?.removeChangeListener(withToken: token)
        }
        
        replicator = nil
    }
}

// MARK: - Sample Document Operations Extension
extension AppServicesSyncManager {
    
    /// Create a new liquor item that will sync to the cloud
    func createLiquorItem(name: String, type: String, price: Double, imageURL: String, quantity: Int = 0) -> String? {
        do {
            let collection = try database.collection(name: collectionName, scope: AppConfig.scopeName) 
                ?? database.createCollection(name: collectionName, scope: AppConfig.scopeName)
            
            let itemId = UUID().uuidString
            let document = MutableDocument(id: itemId)
            
            // Set document properties
            document.setString(itemId, forKey: "id")
            document.setString(name, forKey: "name")
            document.setString(type, forKey: "type")
            document.setDouble(price, forKey: "price")
            document.setString(imageURL, forKey: "imageURL")
            
            // Initialize CRDT counter for quantity
            let quantityCounter = document.crdtCounter(forKey: "quantity", actor: database.deviceUUID ?? "unknown")
            if quantity > 0 {
                quantityCounter.increment(by: UInt(quantity))
            }
            
            // Add metadata
            document.setDate(Date(), forKey: "created_at")
            document.setDate(Date(), forKey: "updated_at")
            document.setString("app_services", forKey: "sync_source")
            document.setString("liquor_item", forKey: "type")
            
            try collection.save(document: document)
            
            print("✅ Created liquor item for App Services sync: \(name) (ID: \(itemId))")
            
            // Trigger immediate sync if enabled
            if isEnabled {
                pushDocumentImmediately(itemId)
            }
            
            return itemId
            
        } catch {
            print("❌ Failed to create liquor item: \(error)")
            return nil
        }
    }
    
    /// Update quantity using CRDT counter (conflict-free)
    func updateLiquorItemQuantity(itemId: String, newQuantity: Int) -> Bool {
        do {
            let collection = try database.collection(name: collectionName, scope: AppConfig.scopeName) 
                ?? database.createCollection(name: collectionName, scope: AppConfig.scopeName)
            
            guard let document = try collection.document(id: itemId)?.toMutable() else {
                print("❌ Document not found: \(itemId)")
                return false
            }
            
            // Get current quantity from CRDT counter
            let quantityCounter = document.crdtCounter(forKey: "quantity", actor: database.deviceUUID ?? "unknown")
            let currentQuantity = quantityCounter.value
            let difference = newQuantity - currentQuantity
            
            // Apply the difference using CRDT operations
            if difference > 0 {
                quantityCounter.increment(by: UInt(difference))
            } else if difference < 0 {
                quantityCounter.decrement(by: UInt(-difference))
            }
            
            // Update metadata
            document.setDate(Date(), forKey: "updated_at")
            document.setString("app_services", forKey: "last_modified_by")
            
            try collection.save(document: document)
            
            print("✅ Updated quantity for \(itemId): \(currentQuantity) → \(newQuantity)")
            
            // Trigger immediate sync if enabled
            if isEnabled {
                pushDocumentImmediately(itemId)
            }
            
            return true
            
        } catch {
            print("❌ Failed to update quantity: \(error)")
            return false
        }
    }
}

File: Android/app/src/main/java/com/example/groceryapplication/AppServicesSyncManager.kt
package com.example.groceryapplication

import android.content.Context
import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.couchbase.lite.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.URI
import java.util.*

// MARK: - App Services Sync State
data class AppServicesSyncState(
    val isConnected: Boolean = false,
    val status: String = "Disconnected",
    val lastSyncTime: Long? = null,
    val documentsInSync: Int = 0,
    val progress: Float = 0f,
    val error: String? = null,
    val totalDocuments: Int = 0,
    val documentsCompleted: Int = 0
)

// MARK: - App Services Sync Manager
class AppServicesSyncManager(
    private val context: Context,
    private val database: Database
) : ViewModel() {
    
    companion object {
        private const val TAG = "AppServicesSync"
    }
    
    // State management
    private val _syncState = MutableStateFlow(AppServicesSyncState())
    val syncState: StateFlow<AppServicesSyncState> = _syncState.asStateFlow()
    
    // Sync control
    var isEnabled by mutableStateOf(false)
        private set
    
    // Couchbase components
    private var replicator: Replicator? = null
    private var replicatorChangeToken: ListenerToken? = null
    private var isSyncActive = false
    
    // NOTE: No init block — sync setup is deferred to after login
    // so that AppConfig.currentStore is set to the correct user's store.
    
    // MARK: - Setup Methods
    
    /**
     * Synchronously sets up the replicator with current AppConfig values.
     * Stops any existing replicator first. Does NOT start replication —
     * call enableAppServices() or setupAndStartSync() for that.
     * Thread-safe: can be called from any thread.
     */
    @Synchronized
    fun setupAppServicesSync() {
        try {
            // Stop and clean up any existing replicator before reconfiguring
            stopSync()
            replicatorChangeToken?.let { replicator?.removeChangeListener(it) }
            replicator = null
            replicatorChangeToken = null
            
            Log.d(TAG, "🔧 Setting up App Services sync configuration...")
            Log.d(TAG, "🔧 Scope: ${AppConfig.scopeName}")
            Log.d(TAG, "🔧 URL: ${AppConfig.syncGatewayURL}")
            Log.d(TAG, "🔧 User: ${AppConfig.username}")
            Log.d(TAG, "🔧 Collections: inventory, profile, orders")
            
            // Get all collections from correct scope (matches Capella structure)
            val inventoryCollection = database.getCollection(AppConfig.COLLECTION_NAME, AppConfig.scopeName)
                ?: database.createCollection(AppConfig.COLLECTION_NAME, AppConfig.scopeName)
            
            val profileCollection = database.getCollection(AppConfig.PROFILE_COLLECTION_NAME, AppConfig.scopeName)
                ?: database.createCollection(AppConfig.PROFILE_COLLECTION_NAME, AppConfig.scopeName)
            
            val ordersCollection = database.getCollection(AppConfig.ORDERS_COLLECTION_NAME, AppConfig.scopeName)
                ?: database.createCollection(AppConfig.ORDERS_COLLECTION_NAME, AppConfig.scopeName)
            
            // Create target endpoint — read dynamically from AppConfig (not frozen companion vals)
            val syncUrl = AppConfig.syncGatewayURL
            Log.d(TAG, "📡 Connecting to: $syncUrl")
            val target = URLEndpoint(URI(syncUrl))
            
            // Create replicator configuration
            val config = ReplicatorConfiguration(target)
            
            // Configure authentication — read dynamically from AppConfig
            config.authenticator = BasicAuthenticator(AppConfig.username, AppConfig.password.toCharArray())
            
            // Configure replication type and behavior
            config.replicatorType = AbstractReplicatorConfiguration.ReplicatorType.PUSH_AND_PULL
            config.isContinuous = AppConfig.SYNC_CONTINUOUS
            config.heartbeat = AppConfig.SYNC_HEARTBEAT.toInt()
            config.maxAttempts = AppConfig.SYNC_MAX_ATTEMPTS
            config.maxAttemptWaitTime = AppConfig.SYNC_MAX_ATTEMPT_WAIT_TIME.toInt()
            
            // Add all collections to replication
            config.addCollection(inventoryCollection, null)
            config.addCollection(profileCollection, null)
            config.addCollection(ordersCollection, null)
            
            // Create replicator
            replicator = Replicator(config)
            
            // Add change listener
            replicatorChangeToken = replicator?.addChangeListener { change ->
                handleReplicationChange(change)
            }
            
            Log.d(TAG, "✅ App Services sync configured successfully")
            updateSyncState { state ->
                state.copy(status = "☁️ Ready to sync")
            }
            
        } catch (e: Exception) {
            Log.e(TAG, "❌ Failed to setup App Services sync", e)
            updateSyncState { state ->
                state.copy(
                    status = "Setup failed",
                    error = e.message
                )
            }
        }
    }
    
    /**
     * Convenience: sets up the replicator with current AppConfig AND immediately starts it.
     * Called from startSyncAfterLogin() to guarantee setup → start ordering.
     */
    @Synchronized
    fun setupAndStartSync() {
        setupAppServicesSync()
        enableAppServices()
    }
    
    // MARK: - Public Sync Control Methods
    @Synchronized
    fun enableAppServices() {
        // Idempotency: skip if replicator is already running.
        // Guarding on isSyncActive (not isEnabled) lets setupAndStartSync()
        // correctly start a freshly reconfigured replicator even when a
        // previous session left isEnabled = true.
        if (isSyncActive) {
            Log.d(TAG, "ℹ️ enableAppServices: sync already active, skipping")
            return
        }

        Log.d(TAG, "🚀 Enabling App Services sync...")
        isEnabled = true
        startSync()

        updateSyncState { state ->
            state.copy(status = "☁️ Starting cloud sync...")
        }
    }

    @Synchronized
    fun disableAppServices() {
        if (!isEnabled) return

        Log.d(TAG, "🛑 Disabling App Services sync...")
        isEnabled = false
        stopSync()

        updateSyncState { state ->
            state.copy(
                status = "☁️ Cloud sync stopped",
                isConnected = false
            )
        }
    }

    @Synchronized
    fun toggleAppServices() {
        if (isEnabled) {
            disableAppServices()
        } else {
            enableAppServices()
        }
    }

    @Synchronized
    private fun startSync() {
        replicator?.let { replicator ->
            if (!isSyncActive) {
                Log.d(TAG, "🌐 Starting App Services replicator...")
                isSyncActive = true
                replicator.start()

                updateSyncState { state ->
                    state.copy(status = "☁️ Connecting to cloud...")
                }
            }
        } ?: run {
            Log.e(TAG, "⚠️ Cannot start sync - replicator not available")
        }
    }

    @Synchronized
    private fun stopSync() {
        if (isSyncActive) {
            Log.d(TAG, "🛑 Stopping App Services replicator...")
            replicator?.stop()
            isSyncActive = false

            updateSyncState { state ->
                state.copy(
                    status = "☁️ Sync stopped",
                    isConnected = false
                )
            }
        }
    }

    @Synchronized
    fun resetSync() {
        Log.d(TAG, "🔄 Resetting App Services sync...")

        stopSync()

        // Restart after a delay, but only if sync is still enabled.
        // The user may have explicitly disabled sync during the delay window.
        viewModelScope.launch {
            kotlinx.coroutines.delay(1000)
            if (isEnabled) startSync()
            else Log.d(TAG, "ℹ️ resetSync: sync was disabled during delay, skipping restart")
        }

        updateSyncState { state ->
            state.copy(
                status = "☁️ Resetting sync...",
                progress = 0f,
                error = null
            )
        }
    }
    
    // MARK: - Replication Event Handling
    private fun handleReplicationChange(change: ReplicatorChange) {
        val status = change.status
        val progress = status.progress
        
        Log.d(TAG, "📊 App Services sync change: ${status.activityLevel} - ${progress.completed}/${progress.total}")
        
        updateSyncState { state ->
            // Update connection status
            val isConnected = (status.activityLevel == ReplicatorActivityLevel.BUSY || 
                             status.activityLevel == ReplicatorActivityLevel.IDLE)
            
            // Update progress
            val progressValue = if (progress.total > 0) {
                progress.completed.toFloat() / progress.total.toFloat()
            } else 0f
            
            // Update status based on activity
            val statusMessage = when (status.activityLevel) {
                ReplicatorActivityLevel.CONNECTING -> "☁️ Connecting to cloud..."
                ReplicatorActivityLevel.BUSY -> "☁️ Syncing... (${progress.completed}/${progress.total})"
                ReplicatorActivityLevel.IDLE -> "☁️ Cloud sync ready"
                ReplicatorActivityLevel.STOPPED -> "☁️ Sync stopped"
                ReplicatorActivityLevel.OFFLINE -> "☁️ Cloud offline"
            }
            
            var newState = state.copy(
                isConnected = isConnected,
                status = statusMessage,
                progress = progressValue,
                documentsCompleted = progress.completed.toInt(),
                totalDocuments = progress.total.toInt(),
                documentsInSync = progress.total.toInt()
            )
            
            // Update last sync time for idle state
            if (status.activityLevel == ReplicatorActivityLevel.IDLE) {
                newState = newState.copy(
                    lastSyncTime = System.currentTimeMillis(),
                    progress = 1f
                )
            }
            
            // Handle errors
            status.error?.let { error ->
                Log.e(TAG, "❌ App Services sync error: $error")
                
                val errorMessage = when ((error as? Exception)?.message) {
                    null -> "Unknown sync error"
                    else -> when {
                        error.message?.contains("network", ignoreCase = true) == true -> 
                            "Network error - will retry"
                        error.message?.contains("auth", ignoreCase = true) == true -> 
                            "Authentication failed"
                        error.message?.contains("forbidden", ignoreCase = true) == true -> 
                            "Access denied"
                        error.message?.contains("not found", ignoreCase = true) == true -> 
                            "Database not found"
                        else -> "Sync error: ${error.message}"
                    }
                }
                
                newState = newState.copy(
                    status = "☁️ $errorMessage",
                    error = error.message,
                    isConnected = false
                )
            } ?: run {
                newState = newState.copy(error = null)
            }
            
            newState
        }
    }
    
    // MARK: - Document Operations
    fun pushDocumentImmediately(documentId: String) {
        if (!isEnabled || replicator == null) {
            Log.w(TAG, "⚠️ Cannot push document - sync not enabled")
            return
        }
        
        Log.d(TAG, "📤 Triggering immediate push for document: $documentId")
        
        // The document will be automatically picked up by the continuous replicator
        if (!isSyncActive) {
            startSync()
        }
    }
    
    fun createGroceryItem(name: String, type: String, price: Double, imageURL: String, quantity: Int = 0): String? {
        return try {
            val collection = database.getCollection(AppConfig.COLLECTION_NAME, AppConfig.scopeName) 
                ?: database.createCollection(AppConfig.COLLECTION_NAME, AppConfig.scopeName)
            
            val itemId = UUID.randomUUID().toString()
            val document = MutableDocument(itemId)
            
            // Set document properties
            document.setString("id", itemId)
            document.setString("name", name)
            document.setString("type", type)
            document.setDouble("price", price)
            document.setString("imageURL", imageURL)
            document.setInt("quantity", quantity)
            document.setString("document_type", "grocery_item")
            
            // Add metadata
            document.setLong("created_at", System.currentTimeMillis() / 1000)
            document.setLong("updated_at", System.currentTimeMillis() / 1000)
            document.setString("sync_source", "app_services")
            
            collection.save(document)
            
            Log.d(TAG, "✅ Created grocery item for App Services sync: $name (ID: $itemId)")
            
            // Trigger immediate sync if enabled
            if (isEnabled) {
                pushDocumentImmediately(itemId)
            }
            
            itemId
            
        } catch (e: Exception) {
            Log.e(TAG, "❌ Failed to create grocery item", e)
            null
        }
    }
    
    fun updateGroceryItemQuantity(itemId: String, newQuantity: Int): Boolean {
        return try {
            val collection = database.getCollection(AppConfig.COLLECTION_NAME, AppConfig.scopeName) 
                ?: database.createCollection(AppConfig.COLLECTION_NAME, AppConfig.scopeName)
            
            val document = collection.getDocument(itemId)?.toMutable() ?: run {
                Log.e(TAG, "❌ Document not found: $itemId")
                return false
            }
            
            // Update quantity and metadata
            document.setInt("quantity", newQuantity)
            document.setLong("updated_at", System.currentTimeMillis() / 1000)
            document.setString("last_modified_by", "app_services")
            
            collection.save(document)
            
            Log.d(TAG, "✅ Updated quantity for $itemId to $newQuantity")
            
            // Trigger immediate sync if enabled
            if (isEnabled) {
                pushDocumentImmediately(itemId)
            }
            
            true
            
        } catch (e: Exception) {
            Log.e(TAG, "❌ Failed to update quantity", e)
            false
        }
    }
    
    // MARK: - DNS Resolution Helper
    private suspend fun resolveHostToIPs(hostname: String): List<String> {
        return withContext(Dispatchers.IO) {
            try {
                val addresses = java.net.InetAddress.getAllByName(hostname)
                val ips = addresses.map { it.hostAddress }.filterNotNull()
                Log.d(TAG, "🔍 Resolved $hostname to IPs: ${ips.joinToString(", ")}")
                ips
            } catch (e: Exception) {
                Log.w(TAG, "❌ Failed to resolve $hostname: ${e.message}")
                emptyList()
            }
        }
    }
    
    // MARK: - Network Connectivity Testing
    private suspend fun testNetworkConnectivity() {
        try {
            val hostname = "orqhtoi5jy2tbev.apps.cloud.couchbase.com"
            Log.d(TAG, "🔍 Testing network connectivity...")
            Log.d(TAG, "🌐 Target hostname: $hostname")
            
            // Test basic internet connectivity first
            withContext(Dispatchers.IO) {
                try {
                    // Test multiple DNS servers to ensure internet is working
                    val dnsServers = listOf("8.8.8.8", "1.1.1.1", "8.8.4.4")
                    var dnsWorking = false
                    
                    for (dns in dnsServers) {
                        try {
                            java.net.InetAddress.getAllByName(dns)
                            Log.d(TAG, "✅ Internet connectivity confirmed ($dns reachable)")
                            dnsWorking = true
                            break
                        } catch (e: Exception) {
                            Log.w(TAG, "⚠️ DNS server $dns not reachable: ${e.message}")
                        }
                    }
                    
                    if (!dnsWorking) {
                        Log.e(TAG, "❌ No DNS servers reachable - network might be down")
                    }
                    
                    // Test general connectivity to couchbase.com
                    try {
                        val couchbaseAddresses = java.net.InetAddress.getAllByName("couchbase.com")
                        Log.d(TAG, "✅ Couchbase.com is reachable (${couchbaseAddresses.size} addresses)")
                    } catch (e: Exception) {
                        Log.w(TAG, "⚠️ Cannot reach couchbase.com: ${e.message}")
                    }
                    
                    // Now test our specific hostname
                    val addresses = java.net.InetAddress.getAllByName(hostname)
                    Log.d(TAG, "✅ DNS resolved to ${addresses.size} addresses:")
                    addresses.forEach { addr ->
                        Log.d(TAG, "   📍 ${addr.hostAddress}")
                    }
                    
                } catch (e: Exception) {
                    Log.e(TAG, "❌ DNS resolution failed for $hostname: ${e.message}")
                    
                    // Try alternative approaches
                    Log.d(TAG, "🔄 Trying alternative connectivity test...")
                    
                    // Skip DNS test and proceed (maybe DNS is slow but connection will work)
                    Log.w(TAG, "⚠️ Proceeding despite DNS test failure - connection might still work")
                    return@withContext
                }
            }
            
            Log.d(TAG, "✅ Network connectivity test passed")
        } catch (e: Exception) {
            Log.e(TAG, "❌ Network connectivity test failed: ${e.message}")
            Log.w(TAG, "⚠️ Continuing anyway - sync might still work")
            
            // Don't throw the exception - let sync attempt proceed
            updateSyncState { current ->
                current.copy(
                    error = "Network test failed, but trying sync anyway: ${e.message}",
                    status = "Testing Connection"
                )
            }
        }
    }
    
    // MARK: - Helper Methods
    private fun updateSyncState(update: (AppServicesSyncState) -> AppServicesSyncState) {
        _syncState.value = update(_syncState.value)
    }
    
    // MARK: - Status Information
    fun getSyncStatusSummary(): String {
        if (!isEnabled) {
            return "App Services sync disabled"
        }
        
        val currentState = _syncState.value
        val baseStatus = currentState.status
        
        return if (currentState.isConnected && currentState.lastSyncTime != null) {
            val time = java.text.SimpleDateFormat("HH:mm:ss", java.util.Locale.getDefault())
                .format(Date(currentState.lastSyncTime))
            "$baseStatus • Last: $time"
        } else {
            baseStatus
        }
    }
    
    fun getDebugInfo(): Map<String, Any> {
        val currentState = _syncState.value
        return mapOf(
            "enabled" to isEnabled,
            "connected" to currentState.isConnected,
            "status" to currentState.status,
            "progress" to currentState.progress,
            "documents_in_sync" to currentState.documentsInSync,
            "last_sync" to (currentState.lastSyncTime ?: 0),
            "error" to (currentState.error ?: "none"),
            "replicator_active" to isSyncActive,
            "sync_gateway_url" to AppConfig.syncGatewayURL,
            "username" to AppConfig.username
        )
    }
    
    // MARK: - Cleanup
    override fun onCleared() {
        super.onCleared()
        Log.d(TAG, "🧹 Cleaning up AppServicesSyncManager...")
        
        stopSync()
        
        replicatorChangeToken?.let { token ->
            replicator?.removeChangeListener(token)
        }
        
        replicator = null
    }
}
